cpp/include/Ice/ServerAuthenticationOptions.h

```diff
@@ -0,0 +1,183 @@
+#    endif
+#    ifdef SECURITY_KERNEL
+#        undef SECURITY_KERNEL
+#    endif
+#    define SECURITY_WIN32 1
+#    include <schannel.h>
+#    include <security.h>
+#    include <sspi.h>
+#    undef SECURITY_WIN32
+#elif defined(__APPLE__)
+#    include <Security/SecureTransport.h>
+#    include <Security/Security.h>
+#else
+#    include <openssl/ssl.h>
+#endif
+
+namespace Ice::SSL
+{
+    /**
+     * The SSL configuration properties for server connections.
+     */
+    struct ServerAuthenticationOptions
+    {
+#if defined(_WIN32)
+        /**
+         * A callback that allows selecting the server credentials based on the server's adapter name.
+         *
+         * @param adapterName The server's adapter name.
+         * @return The server credentials. The credentials must remain valid for the duration of the connection.
+         *
+         * [See Detailed Schannel documentation on Schannel credentials](
+         * https://learn.microsoft.com/en-us/windows/win32/secauthn/acquirecredentialshandle--schannel)
+         */
+        std::function<CredHandle(const std::string& adapterName)> serverCredentialsSelectionCallback;
+
+        /**
+         * A value indicating whether or not the server sends a client certificate request to the client.
+         */
+        bool clientCertificateRequired;
+
+        /**
+         * A callback that allows manually validating the client certificate chain. When the verification callback
+         * returns false, the connection will be aborted with an Ice::SecurityException.
+         *
+         * @param context A CtxtHandle representing the security context associated with the current connection. This
+         * context contains security data relevant for validation, such as the client's certificate chain and cipher
+         * suite.
+         * @param info The IceSSL::ConnectionInfoPtr object that provides additional connection-related data
+         * which might be relevant for contextual certificate validation.
+         * @return true if the certificate chain is valid and the connection should proceed; false if the certificate
+         * chain is invalid and the connection should be aborted. An exception may be thrown to indicate a custom
+         * error during the validation process.
+         *
+         * [See Manually Validating Schannel Credentials for more
+         * details](https://learn.microsoft.com/en-us/windows/win32/secauthn/manually-validating-schannel-credentials).
+         */
+        std::function<bool(CtxtHandle context, const IceSSL::ConnectionInfoPtr& info)>
+            clientCertificateValidationCallback;
+
+#elif defined(__APPLE__)
+        /**
+         * A callback that allows selecting the server's certificate chain based on the server's adapter name.
+         *
+         * @param adapterName The server's adapter name.
+         * @return The server's certificate chain. The certificate chain must remain valid for the duration of the
+         * connection.
```

Comment:
Say I have a long-running server that changes the certificates it uses over time.

How can the application figure out an old cerficate chain is no longer in use by any connection?

Reply:
If you use this callback you cannot really know. It would require an extra connection context parameter with a close callback or something of this sort.

The other option is to use the `sslNewSessionCallback` this works with a per session object, which is managed by the SSL transport. Note the current API doesn't provide this callback for Schannel.